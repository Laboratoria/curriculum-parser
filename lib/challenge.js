import { readdir, readFile, stat } from 'node:fs/promises';
import path from 'node:path';
import {
  detectLangs,
  getTitle,
  parseDirname,
  parseReadmes,
  parser,
  parseTracks,
} from './common.js';

const knownEnvs = ['cjs', 'dom', 'form'];

const dirToTree = (dir) => {
  const recurse = async (current, basedir) => {
    const files = await readdir(current);
    const filtered = files.filter(fname => !fname.endsWith('.md'));

    return Promise.all(filtered.map(async (file) => {
      const stats = await stat(path.join(current, file));
      const relative = path.join(current, file).replace(basedir, '');

      if (stats.isDirectory()) {
        return {
          path: relative,
          children: await recurse(path.join(current, file), basedir),
        };
      }

      return {
        path: relative,
        content: await readFile(path.join(current, file), 'utf8'),
      };
    }));
  };

  return recurse(dir, dir);
};

const flattenTree = tree => tree.reduce(
  (memo, item) => {
    if (typeof item.content === 'string') {
      return { ...memo, [item.path]: item.content };
    }

    return {
      ...memo,
      ...flattenTree(item.children),
    };
  },
  {},
);

const parseForm = async (children) => {
  const questionsHeadingIdx = children.findIndex(node => (
    node.type === 'heading'
    && node.depth === 2
    && node.children?.length
    && node.children[0]?.type === 'text'
    && /(Preguntas|Perguntas)/i.test(node.children[0].value)
  ));

  const nextHeadingIdx = children.slice(questionsHeadingIdx + 1).findIndex(
    ({ type, depth }) => type === 'heading' && depth === 2,
  );

  const questionsNodes = children.slice(
    questionsHeadingIdx + 1,
    nextHeadingIdx === -1 ? undefined : questionsHeadingIdx + nextHeadingIdx,
  );

  const questions = questionsNodes.reduce(
    (memo, node) => {
      if (node.type === 'heading' && node.depth === 3) {
        return memo.concat({
          title: node.children[0].value,
          body: [],
        });
      }

      const current = memo[memo.length - 1];

      if (!current) {
        return memo;
      }

      if (
        node.type === 'heading'
        && node.depth === 4
        && node.children?.length
        && node.children[0]?.type === 'text'
        && /(Opciones|Opções)/i.test(node.children[0].value)
      ) {
        Object.assign(current, { options: [] });
        return memo;
      }

      if (
        node.type === 'heading'
        && node.depth === 4
        && node.children?.length
        && node.children[0]?.type === 'text'
        && /(Solución|Solução)/i.test(node.children[0].value)
      ) {
        Object.assign(current, { solution: {} });
        return memo;
      }

      if (current?.options && !current.solution && node.type === 'list' && node.ordered) {
        Object.assign(current, {
          options: node.children.map(child => child.children),
        });
        return memo;
      }

      if (current.solution && node.type === 'list' && !node.ordered) {
        const items = node.children.reduce(
          (prev, child) => {
            if (child.type !== 'listItem') {
              return prev;
            }
            // First paragraph in the list item has the item name and sometimes
            // also its value.
            const firstP = child.children[0];
            // The first child of the firstP is expected to be a <code> tag
            // qith the item name.
            if (firstP.children[0].type !== 'inlineCode') {
              return prev;
            }
            const name = firstP.children[0].value;
            const value = (
              name !== 'validate'
                ? firstP.children[1].value.trim().replace(/^:\s?/, '')
                : child.children[1]?.value
            );

            if (name === 'value' && current.options) {
              const val = value.split(',').map(str => parseInt(str, 10) - 1);
              return {
                ...prev,
                [name]: val.length > 1 ? val : val[0],
              };
            }

            return {
              ...prev,
              [name]: (
                ['required', 'multiline'].includes(name)
                  ? value !== 'false'
                  : value
              ),
            };
          },
          { required: true },
        );
        Object.assign(current.solution, items);
        return memo;
      }

      Object.assign(current, { body: current.body.concat(node) });
      return memo;
    },
    [],
  );

  return {
    body: parser.stringify(await parser.run({
      type: 'root',
      children: children.slice(0, questionsHeadingIdx),
    })),
    questions: await Promise.all(questions.map(async question => ({
      ...question,
      body: parser.stringify(await parser.run({
        type: 'root',
        children: question.body,
      })),
      ...(question.options && {
        options: await Promise.all(question.options.map(
          async opt => parser.stringify(await parser.run({
            type: 'root',
            children: opt,
          })),
        )),
      }),
    }))),
  };
};

export const parseChallenge = async (dir, opts, pkg) => {
  const { prefix, slug } = parseDirname(dir);
  const langs = await detectLangs(dir);

  const { parsedLocales, meta } = await parseReadmes(
    dir,
    langs,
    'challenge',
    async (rootNode, localizedMeta) => {
      const { env } = localizedMeta || {};
      const children = rootNode.children.slice(1);

      return {
        title: getTitle(rootNode),
        ...(
          env === 'form'
            ? await parseForm(children)
            : {
              body: parser.stringify(await parser.run({
                type: 'root',
                children,
              })),
            }
        ),
      };
    },
  );

  if (!knownEnvs.includes(meta.env)) {
    throw new Error(`Unknown challenge env "${meta.env}"`);
  }

  // TODO: Validate files (boilerplate, tests and solution)
  const files = flattenTree(await dirToTree(dir));

  const { track, tracks } = parseTracks(meta);

  return {
    slug,
    prefix,
    repo: opts?.repo,
    path: path.relative(process.cwd(), dir),
    version: opts?.version,
    parserVersion: pkg?.version,
    createdAt: new Date(),
    track,
    tracks,
    env: meta.env,
    intl: langs.reduce(
      (memo, lang, idx) => ({
        ...memo,
        [lang]: parsedLocales[idx],
      }),
      {},
    ),
    files,
  };
};
