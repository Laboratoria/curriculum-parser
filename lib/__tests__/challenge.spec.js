import { describe, expect, it, vi } from 'vitest';
import { resolveFixturePath } from './helpers.js';
import { parseChallenge } from '../challenge.js';

vi.mock('sharp');

describe('parseChallenge', () => {
  it('throws when unknown env', () => {
    const dir = resolveFixturePath('challenge-unknown-env');
    return expect(parseChallenge(dir, {}, {}))
      .rejects.toThrow('Unknown challenge env "foo"');
  });

  it('parses cjs challenge', async () => {
    const dir = resolveFixturePath('challenge-cjs');
    const parsed = await parseChallenge(dir, {}, {});
    expect(parsed.env).toBe('cjs');
  });

  it('parses form challenge', async () => {
    const dir = resolveFixturePath('challenge-form');
    const parsed = await parseChallenge(dir, {}, {});
    expect(parsed.env).toBe('form');
    expect(parsed.track).toBe('data');
    expect(parsed.intl.es.questions).toEqual([
      {
        title: 'What is the meaning of the universe?',
        body: '<p>Texto complementarion de la pregunta.</p>',
        solution: '42',
      },
      {
        title: 'A question with multiple choices',
        body: '',
        options: ['<p>A</p>', '<p>B</p>', '<p>C</p>'],
        solution: 1,
      },
      {
        title: 'Cuáles de las siguientes crees que son correctas?',
        body: '',
        options: ['<p>aaa</p>', '<p>bbb</p>', '<p>ccc</p>'],
        solution: [0, 1],
      },
      {
        title: 'Anything else to say?',
        body: '<p>¿Alguna cosa más que agregar?</p>',
      },
    ]);
  });
});
