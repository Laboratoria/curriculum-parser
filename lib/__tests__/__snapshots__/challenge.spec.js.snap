// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseChallenge > ignores questions with no heading 1`] = `
[
  {
    "body": "",
    "title": "A question",
  },
]
`;

exports[`parseChallenge > parses form challenge 1`] = `
[
  {
    "body": "<p>Texto complementarion de la pregunta.</p>",
    "solution": {
      "required": true,
      "value": "42",
    },
    "title": "What is the meaning of the universe?",
  },
  {
    "body": "",
    "options": [
      "<p>A</p>",
      "<p>B</p>",
      "<p>C</p>",
    ],
    "solution": {
      "required": true,
      "value": 1,
    },
    "title": "A question with multiple choices",
  },
  {
    "body": "",
    "options": [
      "<p>A</p>",
      "<p>B</p>",
      "<p>C</p>",
    ],
    "solution": {
      "required": false,
      "value": 1,
    },
    "title": "An optional question with multiple choices",
  },
  {
    "body": "",
    "options": [
      "<p>aaa</p>",
      "<p>bbb</p>",
      "<p>ccc</p>",
    ],
    "solution": {
      "required": true,
      "value": [
        0,
        1,
      ],
    },
    "title": "Cuáles de las siguientes crees que son correctas?",
  },
  {
    "body": "<p>¿Alguna cosa más que agregar?</p>",
    "solution": {
      "required": false,
    },
    "title": "Anything else to say?",
  },
  {
    "body": "",
    "solution": {
      "required": true,
    },
    "title": "A required question without validation",
  },
  {
    "body": "",
    "solution": {
      "required": true,
      "validate": "(value) => {
  if (!value.startWith('https://blah.blah/')) {
    return 'Respuesta debe comenzr con https://blah.blah/';
  }
}",
    },
    "title": "A question with custom validation",
  },
]
`;

exports[`parseChallenge > parses form challenge with multiline question 1`] = `
[
  {
    "body": "<p>Texto complementarion de la pregunta.</p>",
    "solution": {
      "multiline": true,
      "required": true,
    },
    "title": "What is the meaning of the universe?",
  },
]
`;
